### PR TITLE
Implementing Putmask

### PR DESCRIPTION
**Pull request by ipdemes**

### CODE REVIEW

**Comment by ipdemes** [comment]
This PR implements logic for putmask funtion.
It will also optimize case of Advanced Indexing operation like `a[bool_indices]=scalar` by calling `putmask` instead of doing Sparse Indirect Copy. 
The last commit is a replacement to PR https://github.com/nv-legate/cunumeric/pull/639

File: cunumeric/module.py [hunk 1]
```diff
@@ -3506,6 +3506,46 @@ def put(
...
+    if not a.shape == mask.shape:
+        raise ValueError("mask and data must be the same size")
+
+    mask = mask._warn_and_convert(np.dtype(bool))
+
+    if a.dtype != values.dtype:
+        values = values._warn_and_convert(a.dtype)
+    if values.shape != a.shape and values.size != 1:
+        values = values._wrap(a.size)
+        values = values.reshape(a.shape)
```
Comment (magnatelee) [inline cunumeric/module.py:3545]: what's the benefit of doing this over tiling followed by slicing? if I understand it correctly, this code creates three temporary stores, one for indices, one for the copy output, and one from the final reshaping. In comparison, the tiling would materialize only one temporary. Plus, tiling is a much simpler operation than the gather copy you do in `_wrap`. Please consider rewriting this with slicing.

File: src/cunumeric/execution_policy/indexing/bool_mask.cuh [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    bool_mask_dense_kernel<<<blocks, THREADS_PER_BLOCK, 1, stream>>>(
+      volume, maskptr, std::forward<KERNEL>(kernel));
+
+    CHECK_CUDA_STREAM(stream);
+  }
+};
+
+template <>
+struct BoolMaskPolicy<VariantKind::GPU, false> {
```
Comment (magnatelee) [inline src/cunumeric/execution_policy/indexing/bool_mask.cuh:63]: does this need to be specific to boolean masks? this policy cannot be used without materializing a boolean mask in a case where, for example, the kernel needs to run when the input is not NaN. it'd be more general if this policy took a predicate that evaluates to a boolean value on each point.

File: cunumeric/deferred.py [hunk 3]
```diff
@@ -1740,6 +1766,28 @@ def put(self, indices: Any, values: Any, check_bounds: bool) -> None:
         if self_tmp is not self:
             self.copy(self_tmp, deep=True)
 
+    @auto_convert("mask", "values")
+    def putmask(self, mask: Any, values: Any) -> None:
+        is_scalar_value = False
+        if values.shape != self.shape and values.size == 1:
+            if values.shape == ():
+                values = values._convert_future_to_regionfield(True)
```
Comment (magnatelee) [inline cunumeric/deferred.py:1774]: is there a reason that you do this? can't you just promote this scalar store to match the shape of `self`?

File: cunumeric/deferred.py [hunk 4]
```diff
@@ -1740,6 +1766,28 @@ def put(self, indices: Any, values: Any, check_bounds: bool) -> None:
...
+            is_scalar_value = True
+
+        task = self.context.create_task(CuNumericOpCode.PUTMASK)
+        task.add_input(self.base)
+        task.add_input(mask.base)
+        task.add_input(values.base)
+        task.add_output(self.base)
+        task.add_scalar_arg(is_scalar_value, bool)  # value is scalar
```
Comment (magnatelee) [inline cunumeric/deferred.py:1782]: again, is there a reason that you don't do numpy broadcasting on the scalar value? that would make the code a lot simpler.

File: cunumeric/deferred.py [hunk 5]
```diff
@@ -1740,6 +1766,28 @@ def put(self, indices: Any, values: Any, check_bounds: bool) -> None:
...
+        task.add_alignment(self.base, mask.base)
+        if is_scalar_value:
+            task.add_broadcast(values.base)
+        else:
+            task.add_alignment(self.base, values.base)
+        task.add_broadcast(self.base, axes=range(1, self.ndim))
```
Comment (magnatelee) [inline cunumeric/deferred.py:1788]: I don't understand this broadcast constraint... why is this necessary?

File: src/cunumeric/index/putmask_template.inl [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+#include "cunumeric/index/putmask.h"
+#include "cunumeric/pitches.h"
+#include "cunumeric/execution_policy/indexing/bool_mask.h"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+template <VariantKind KIND, LegateTypeCode CODE, int DIM, int VDIM, bool SCALAR_VALUE = false>
```
Comment (magnatelee) [inline src/cunumeric/index/putmask_template.inl:30]: like I said in the other comment, if you do numpy broadcasting on the values, you can remove the `SCALAR_VALUE == true` case.

File: cunumeric/module.py [hunk 7]
```diff
@@ -3506,6 +3506,46 @@ def put(
...
+    Multiple GPUs, Multiple CPUs
+    """
+    if not a.shape == mask.shape:
+        raise ValueError("mask and data must be the same size")
+
+    mask = mask._warn_and_convert(np.dtype(bool))
+
+    if a.dtype != values.dtype:
+        values = values._warn_and_convert(a.dtype)
+    if values.shape != a.shape and values.size != 1:
```
Comment (magnatelee) [inline cunumeric/module.py:3543]: the first condition should be `values.size != size`. Here's an example you can handle without wrapping or tiling but simply using numpy broadcasting:

```
import numpy as np
a = np.full((3, 3), 3)
np.putmask(a, np.full((3, 3), True), np.full(3, 10))
```

Comment (magnatelee) [reply cunumeric/module.py:3543]: unless stated otherwise, it's safe to assume that all related array arguments follow the numpy broadcasting semantics.

File: tests/integration/test_putmask.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    assert np.array_equal(np_arr, num_arr)
+
+    # val is the same shape:
+    np_val = np_arr * 10
+    num_val = num_arr * 10
+    np.putmask(np_arr, np_mask, np_val)
+    num.putmask(num_arr, num_mask, num_val)
+    assert np.array_equal(np_arr, num_arr)
+
+    # val is different shape
```
Comment (magnatelee) [inline tests/integration/test_putmask.py:134]: you need to subdivide this test into two: 1) when the shapes are different but the sizes are the same, 2) when both the shapes and sizes are different.

**Review by magnatelee — CHANGES_REQUESTED** [review]

File: cunumeric/module.py (hunk 1 above)
Comment (ipdemes) [reply cunumeric/module.py:3545]: @magnatelee , consider this example:
```
import numpy as np

arr = np.array([[[1, 2, 3, 4, 5]]])
ind = np.array([[[True,True, False, False,True]]])
v = np.array([[11,12],[13,14]])
np.putmask(arr,ind,v)
print(arr)
```
this will output:
```
[[[11 12  3  4 11]]]
```
In this case shape of v is 2d and shape of mask is 3d. In this case we could do following instead of `_wrap`:
```
v_flat = v.ravel()
old_len = v.size
new_len = a.size
reps = (new_len + old_len - 1) // old_len
v_new = v_flat.tile(reps) if reps > 1 else v_flat
v_new = v_new[:new_len].copy()
v_new = v.reshape(a.shape)
```
This approach would avoid gather/scatter copy, however there is a risk of allocating too much memory for v_new if `new_len -old_len` is a small value and `old_len` is very large

Should I anyway go with this `tile` approach?

Comment (magnatelee) [reply cunumeric/module.py:3545]: First of all, I think it's worth adding an optional out or shape parameter to `tile` to express a tiling up to a shape. Then we don't need to worry about `v_new` getting too big. Second, an input array to `tile` being too big is already a problem and we can potentially fix that by falling back to `cunumeric.block. Or we could use `_wrap` in that case. In either case, I feel we can't get away with it and some heuristic based on the size is inevitable in choosing the right implementation.

Comment (magnatelee) [reply cunumeric/module.py:3545]: that said, I guess we should address it in a follow-on PR and not this one, so maybe we just keep the current code. I expect the common case to be those where the `values` array can be easily broadcasted to match the target's shape, so we should focus on making them run fast.

Comment (ipdemes) [reply cunumeric/module.py:3545]: sounds good!

File: src/cunumeric/execution_policy/indexing/bool_mask.cuh (hunk 2 above)
Comment (ipdemes) [reply src/cunumeric/execution_policy/indexing/bool_mask.cuh:63]: Makes sense. I will make it more general and call it `parallel_loop`, probably.

File: cunumeric/deferred.py (hunk 4 above)
Comment (ipdemes) [reply cunumeric/deferred.py]: makes sense. replaced the logic with broadcasting

File: cunumeric/deferred.py (hunk 5 above)
Comment (ipdemes) [reply cunumeric/deferred.py]: removed. I don't need it anymore

File: cunumeric/deferred.py (hunk 3 above)
Comment (ipdemes) [reply cunumeric/deferred.py]: replaced this with broadcasting

File: cunumeric/module.py (hunk 7 above)
Comment (ipdemes) [reply cunumeric/module.py]: I adding the check if arrays can be broadcasted. We will cal `wrap` only for the case when they can't

File: cunumeric/deferred.py [hunk 9]
```diff
@@ -583,61 +586,79 @@ def _create_indexing_array(
...
+            # and avoid calling Copy
+            has_set_value = set_value is not None and set_value.size == 1
+            if has_set_value:
+                mask = DeferredArray(
+                    self.runtime,
+                    base=key_store,
+                    dtype=self.dtype,
+                )
+                rhs.putmask(mask, set_value)
+                return False, rhs, rhs, self
```
Comment (magnatelee) [inline cunumeric/deferred.py:601]: I'm not a big fan of this code. the name of the function is `_create_indexing_array`, but this code makes it do more than just creating indexing arrays. and in fact, the function was already complicated enough to warrant a refactoring, which I was thinking of doing at some point. I guess I'm fine with accepting this edit for this PR, but keep in mind that `_create_indexing_array` can use some refactoring for better maintainability.

File: cunumeric/module.py [hunk 10]
```diff
@@ -3506,6 +3506,58 @@ def put(
     a.put(indices=indices, values=values, mode=mode)
 
 
+def _can_broadcast_to_shape(self: ndarray, shape: NdShape) -> bool:
```
Comment (magnatelee) [inline cunumeric/module.py:3509]: can't we just use `np.broadcast_shapes` instead of this custom code?

Comment (ipdemes) [reply cunumeric/module.py:3509]: sure, will use it instead with `try\ except"

**Review by magnatelee — APPROVED** [review]
